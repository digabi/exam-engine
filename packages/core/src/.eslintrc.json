{
  "rules": {
    "no-restricted-imports": [
      "error",
      {
        "paths": [
          {
            "name": "lodash",
            "message": "Please use import * as _ from 'lodash-es' in the core module for better tree shaking."
          },
          {
            "name": "react-i18next",
            "importNames": ["useTranslation", "Translation"],
            "message": "Please use the `useExamTranslation` wrapper instead."
          }
        ]
      }
    ]
  }
}
