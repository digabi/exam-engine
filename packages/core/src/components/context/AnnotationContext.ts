import React from 'react'
import { ExamAnnotation } from '../../types/Score'

//type Annotation = ExamAnnotation | TextAnnotation
export interface AnnotationContextType {
  annotations: Record<string, ExamAnnotation[]>
  onClickAnnotation: (e: React.MouseEvent<HTMLElement, MouseEvent>, a: ExamAnnotation) => void
  onSaveAnnotation: (a: ExamAnnotation, key: string) => void
  setNewAnnotation: (a: ExamAnnotation | null) => void
  newAnnotation: ExamAnnotation | null
}

export const AnnotationContext = React.createContext({} as AnnotationContextType)
