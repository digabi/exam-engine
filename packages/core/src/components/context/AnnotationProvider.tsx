import * as _ from 'lodash-es'
import React from 'react'
import { NewExamAnnotation, RenderableAnnotation } from '../../types/Score'
import { AnnotationProps } from '../exam/Exam'
import { onMouseDownForAnnotation } from '../grading/examAnnotationUtils'
import { AnnotationPopup } from '../shared/AnnotationPopup'

interface Props {
  children: React.ReactNode
}

export interface RenderableAnnotationProps {
  annotations: Record<string, RenderableAnnotation[]>
  onClickAnnotation?: (e: React.MouseEvent<HTMLElement, MouseEvent>, annotation: RenderableAnnotation) => void
  onSaveAnnotation?: (annotation: NewExamAnnotation, comment: string) => void
}

export interface AnnotationContextType extends RenderableAnnotationProps {
  newAnnotation: NewExamAnnotation | null
  setNewAnnotation: (a: NewExamAnnotation | null) => void
  newAnnotationRef: HTMLElement | undefined
  setNewAnnotationRef: (ref: HTMLElement | undefined) => void
}

export const AnnotationContext = React.createContext({} as AnnotationContextType)

export const AnnotationProvider = ({
  children,
  annotations,
  onClickAnnotation,
  onSaveAnnotation
}: Props & AnnotationProps) => {
  const [newAnnotation, setNewAnnotation] = React.useState<NewExamAnnotation | null>(null)
  const [newAnnotationRef, setNewAnnotationRef] = React.useState<HTMLElement>()

  const mouseUpCallback = (annotation: NewExamAnnotation) => {
    setNewAnnotation(annotation)
  }

  function onMouseDown(e: React.MouseEvent) {
    const target = e.target as Element
    const clickIsInPopup = !!target.closest('.annotation-popup')
    if (!clickIsInPopup) {
      setNewAnnotation(null)
    }
    onMouseDownForAnnotation(e, mouseUpCallback)
  }

  if (!onClickAnnotation && !onSaveAnnotation) {
    return children
  }

  const annotationPartsToRenderableAnnotations: Record<string, RenderableAnnotation[]> = _.groupBy(
    annotations?.flatMap(a =>
      a.annotationParts.map((p, index, arr) => {
        const isLastChild = index === arr.length - 1
        return {
          ...p,
          annotationId: a.annotationId,
          markNumber: isLastChild ? a.markNumber : undefined,
          hidden: a.hidden
        }
      })
    ) || [],
    'annotationAnchor'
  )

  return (
    <AnnotationContext.Provider
      value={{
        annotations: annotationPartsToRenderableAnnotations,
        onClickAnnotation,
        onSaveAnnotation,
        newAnnotation,
        setNewAnnotation,
        newAnnotationRef,
        setNewAnnotationRef
      }}
    >
      <span onMouseDown={onMouseDown}>
        <AnnotationPopup />
        {children}
      </span>
    </AnnotationContext.Provider>
  )
}
