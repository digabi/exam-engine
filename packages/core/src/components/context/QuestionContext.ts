import React from 'react'
import { findChildElement, getNumericAttribute, parentElements, queryAll } from '../../dom-utils'
import { withContext } from './withContext'
import { ExamComponentProps } from '../../createRenderChildNodes'

export interface QuestionContext {
  answers: Element[]
  displayNumber: string
  hasExternalMaterial: boolean
  maxAnswers?: number
  maxScore: number
  level: number
  childQuestions: Element[]
}

export const QuestionContext = React.createContext<QuestionContext>({} as QuestionContext)

export const withQuestionContext = withContext<QuestionContext, ExamComponentProps>(QuestionContext, ({ element }) => {
  const childQuestions = queryAll(element, 'question', false)
  const answers = childQuestions.length
    ? []
    : queryAll(element, ['text-answer', 'scored-text-answer', 'dropdown-answer', 'choice-answer'], false)

  return {
    answers,
    displayNumber: element.getAttribute('display-number')!,
    hasExternalMaterial: findChildElement(element, 'external-material') != null,
    maxAnswers: getNumericAttribute(element, 'max-answers'),
    maxScore: getNumericAttribute(element, 'max-score')!,
    level: parentElements(element, 'question').length,
    childQuestions,
  }
})
