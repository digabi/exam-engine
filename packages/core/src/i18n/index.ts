import i18n from 'i18next'
import * as _ from 'lodash-es'
import { A_E } from './A_E'
import { BA } from './BA'
import { CA } from './CA'
import { EA } from './EA'
import { FA } from './FA'
import { fi_FI } from './fi-FI'
import { GC } from './GC'
import { I } from './I'
import { M } from './M'
import { O_E } from './O_E'
import { PA } from './PA'
import { SA } from './SA'
import { sv_FI } from './sv-FI'
import { TC } from './TC'
import { VA } from './VA'
import { Z } from './Z'

const examSpecificTranslations = {
  A_E,
  BA,
  BB: BA,
  CA,
  CB: CA,
  EA,
  EC: EA,
  FA,
  FC: FA,
  GC,
  I,
  M,
  N: M,
  O_E,
  SA,
  SC: SA,
  PA,
  PC: PA,
  TC,
  VA,
  VC: VA,
  Z,
}

const resources = {
  'fi-FI': {
    translation: fi_FI,
    ...examSpecificTranslations,
  },
  'sv-FI': {
    translation: sv_FI,
    ...examSpecificTranslations,
  },
}

export function initI18n(language: string, examCode: string | null, dayCode: string | null): typeof i18n {
  const namespace = examCode ? examCode + (dayCode ? `_${dayCode}` : '') : 'translation'

  return i18n.createInstance(
    {
      resources,
      lng: language,
      fallbackLng: 'fi-FI',
      debug: false,
      defaultNS: namespace,
      fallbackNS: 'translation',
      interpolation: {
        escapeValue: false,
        format: (value, format) => {
          switch (format) {
            case 'range': {
              // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
              const [start, end] = value
              // eslint-disable-next-line @typescript-eslint/no-unsafe-return,@typescript-eslint/restrict-template-expressions
              return start == null ? end : end == null ? start : start === end ? start : `${start}â€“${end}`
            }
            case 'first':
              return _.first(value)
            case 'last':
              return _.last(value)
            default:
              // eslint-disable-next-line @typescript-eslint/no-unsafe-return
              return value
          }
        },
      },
    },
    _.noop
  )
}

export const changeLanguage = (_i18n: typeof i18n, language: string) => (): void => {
  if (_i18n.language !== language) {
    _i18n.changeLanguage(language).then(_.noop).catch(_.noop)
  }
}

export default i18n
