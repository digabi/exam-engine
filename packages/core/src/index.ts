import '../src/css/main.less'
import Attachments from './components/attachments/Attachments'
import Exam from './components/exam/Exam'
import GradingInstructions from './components/grading-instructions/GradingInstructions'
import Results from './components/results/Results'
import parseExam from './parser/parseExam'

export { Attachments, Exam, GradingInstructions, Results, parseExam }
export { ExamComponentProps } from './createRenderChildNodes'
export * from './types/ExamAnswer'
export * from './types/ExamServerAPI'
export * from './types/GradingStructure'
export * from './types/Score'

export interface ExamBundle {
  Attachments: typeof Attachments
  Exam: typeof Exam
  parseExam: typeof parseExam
  Results: typeof Results
}
