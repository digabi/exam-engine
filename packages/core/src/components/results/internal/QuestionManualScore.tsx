import classNames from 'classnames'
import React, { useContext } from 'react'
import { CensoringScore, InspectionScore, PregradingScore, Score } from '../../../index'
import { useExamTranslation } from '../../../i18n'
import { QuestionContext } from '../../context/QuestionContext'
import ResultsExamQuestionScoresContainer from './QuestionScoresContainer'

export interface QuestionManualScoreProps {
  scores?: Score
  maxScore?: number
  displayNumber?: string
  multilineAnswer?: boolean
}

interface NormalizedScore {
  score: number
  shortCode: string
  type: string
}

function QuestionManualScore({ scores, maxScore, displayNumber, multilineAnswer }: QuestionManualScoreProps) {
  const { answers } = useContext(QuestionContext)
  const containerProps = { answers, displayNumber, multilineAnswer }
  const shortCode = scores?.censoring?.nonAnswerDetails?.shortCode

  return (
    <ResultsExamQuestionScoresContainer {...containerProps}>
      {scores?.pregrading?.score == null && !scores?.censoring?.scores?.length ? (
        <NoPregrading {...{ maxScore }} />
      ) : (
        <>
          {renderNormalizedScores(scores, maxScore)}
          {shortCode && <NonAnswer shortCode={shortCode} />}
        </>
      )}
    </ResultsExamQuestionScoresContainer>
  )
}

function renderNormalizedScores(scores: Score, maxScore?: number) {
  const normalizedScores = [
    scores.inspection && normalizeInspectionScore(scores.inspection),
    ...(scores.censoring ? normalizeCensoringScores(scores.censoring) : []),
    scores.pregrading && normalizePregradingScore(scores.pregrading),
  ].filter(Boolean) as NormalizedScore[]

  return normalizedScores.map((score, i) => <ScoreRow key={i} {...score} latest={i === 0} maxScore={maxScore} />)
}

interface NoPregradingProps {
  maxScore?: number
}

function NoPregrading({ maxScore }: NoPregradingProps) {
  const { t } = useExamTranslation()
  return (
    <>
      <span className="e-result-scorecount-empty" />
      {maxScore && ` / ${maxScore} `}
      {t('points')}
    </>
  )
}

interface NonAnswerProps {
  shortCode: string
}

function NonAnswer({ shortCode }: NonAnswerProps) {
  return (
    <div className="e-color-darkgrey e-columns e-columns--center-v">
      <span className="e-font-size-xxxl e-light e-mrg-r-1">Ã—</span>
      <span>{shortCode}</span>
    </div>
  )
}

interface ScoreRowProps {
  maxScore?: number
  latest: boolean
}

function ScoreRow({ score, shortCode, type, maxScore, latest }: ScoreRowProps & NormalizedScore) {
  return (
    <div className={latest ? 'e-color-black' : 'e-color-darkgrey e-font-size-xs'}>
      <ScoreColumn className={classNames('e-nowrap', { 'e-font-size-m': latest })}>
        {latest ? <b>{score}</b> : score}
        {latest && maxScore ? ` / ${maxScore}` : ''}
        {` p.`}
      </ScoreColumn>
      <ScoreColumn className="e-result-scorecount-shortcode">{shortCode}</ScoreColumn>
      <ScoreColumn className="e-mrg-r-0">{type}</ScoreColumn>
    </div>
  )
}

interface ScoreColumnProps {
  children?: React.ReactNode
  className?: string
}

function ScoreColumn({ className, children }: ScoreColumnProps) {
  return <span className={classNames('e-mrg-r-1 e-nowrap', className)}>{children}</span>
}

function normalizePregradingScore({ score }: PregradingScore): NormalizedScore | null {
  return score !== undefined ? { score, shortCode: '', type: 'va' } : null
}

function normalizeCensoringScores(score: CensoringScore): NormalizedScore[] {
  return score.scores.map((s, i) => ({
    score: s.score,
    shortCode: s.shortCode || '',
    type: `${score.scores.length - i}.s`,
  }))
}

function normalizeInspectionScore(score: InspectionScore): NormalizedScore {
  return { score: score.score, shortCode: score.shortCodes ? score.shortCodes.join(', ') : '', type: 'ta' }
}

export default React.memo(QuestionManualScore)
