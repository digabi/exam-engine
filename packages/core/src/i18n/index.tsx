import i18n, { PostProcessorModule, TOptions } from 'i18next'
import * as _ from 'lodash-es'
import React from 'react'
// eslint-disable-next-line no-restricted-imports
import { TFunction, useTranslation } from 'react-i18next'
import { A_E } from './A_E'
import { BA } from './BA'
import { CA } from './CA'
import { EA } from './EA'
import { FA } from './FA'
import { fi_FI } from './fi-FI'
import { GC } from './GC'
import { I } from './I'
import { M } from './M'
import { O_E } from './O_E'
import { PA } from './PA'
import { SA } from './SA'
import { sv_FI } from './sv-FI'
import { TC } from './TC'
import { VA } from './VA'
import { Z } from './Z'

const examSpecificTranslations = {
  A_E,
  BA,
  BB: BA,
  CA,
  CB: CA,
  EA,
  EC: EA,
  FA,
  FC: FA,
  GC,
  I,
  M,
  N: M,
  O_E,
  SA,
  SC: SA,
  PA,
  PC: PA,
  TC,
  VA,
  VC: VA,
  Z,
}

const resources = {
  'fi-FI': {
    translation: fi_FI,
    ...examSpecificTranslations,
  },
  'sv-FI': {
    translation: sv_FI,
    ...examSpecificTranslations,
  },
}

declare module 'react-i18next' {
  type ExamEngineI18nResources = { translation: typeof fi_FI }

  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface Resources extends ExamEngineI18nResources {}
}

export function initI18n(language: string, examCode?: string, dayCode?: string): typeof i18n {
  const namespace = examCode ? examCode + (dayCode ? `_${dayCode}` : '') : 'translation'

  return i18n
    .use<PostProcessorModule>({
      type: 'postProcessor',
      name: 'lowercase',
      process: (value, _key, options) => value.toLocaleLowerCase(options.lng),
    })
    .createInstance(
      {
        compatibilityJSON: 'v3',
        resources,
        lng: language,
        fallbackLng: 'fi-FI',
        debug: false,
        defaultNS: namespace,
        fallbackNS: 'translation',
        interpolation: {
          escapeValue: false,
          format: (value, format) => {
            const val = value as [string, string]
            switch (format) {
              case 'range': {
                const [start, end] = val
                return start == null ? end : end == null ? start : start === end ? start : `${start}–${end}`
              }
              case 'first':
                return _.first(val)
              case 'last':
                return _.last(val)
              default:
                // eslint-disable-next-line @typescript-eslint/no-unsafe-return
                return value
            }
          },
        },
      },
      _.noop
    )
}

export function useExamTranslation() {
  const { i18n, t: originalT } = useTranslation()
  type Translator = {
    (key: string | string[], options?: TOptions): string | JSX.Element
    raw: TFunction
  }
  const t: Translator = (key, options) => translate(i18n, key, options)
  t.raw = originalT
  return { i18n, t }
}

/**
 * Custom wrapper for i18n.t that wraps the translated text in a span element
 * containing the correct lang attribute, if necessary.
 *
 * If you want to disable the wrapping, you can use `t.raw(…)`.
 */
function translate(instance: typeof i18n, key: string | string[], options?: TOptions): string | JSX.Element {
  const translation = instance.t(key, options)

  const namespace = instance.options.defaultNS!.toString()
  const examHasCustomTranslations = namespace !== 'translation' && namespace in examSpecificTranslations
  const hasCustomTranslationForKey = examHasCustomTranslations && hasTranslation(instance, namespace, key)
  return hasCustomTranslationForKey || !examHasCustomTranslations ? (
    translation
  ) : (
    <span lang={instance.language}>{translation}</span>
  )
}

/**
 * A hacky way to check if we have a translation in the specified namespace.
 * Ideally, we'd probably want to use `i18n.exists()` for this purpose, but I
 * couldn't find a way to disable the fallback mechanism when using it.
 */
function hasTranslation(instance: typeof i18n, namespace: string, key: string | string[]): boolean {
  return Array.isArray(key)
    ? key.some((k) => instance.getResource(instance.language, namespace, k) != null)
    : instance.getResource(instance.language, namespace, key) != null
}

export const changeLanguage = (_i18n: typeof i18n, language: string) => (): void => {
  if (_i18n.language !== language) {
    _i18n.changeLanguage(language).then(_.noop).catch(_.noop)
  }
}

export default i18n
