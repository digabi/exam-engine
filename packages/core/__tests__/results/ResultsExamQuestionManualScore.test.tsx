import * as i18next from 'i18next'
import _ from 'lodash'
import React from 'react'
import { I18nextProvider } from 'react-i18next'
import { create } from 'react-test-renderer'
import { Score } from '../../src'
import { QuestionContext } from '../../src/components/context/QuestionContext'
import QuestionManualScore, {
  QuestionManualScoreProps,
} from '../../src/components/results/internal/QuestionManualScore'
import { initI18n } from '../../src/i18n'

const defaultProps = {
  maxScore: 6,
  displayNumber: '1',
}

const defaultScores: Score = {
  questionId: 1,
  answerId: 1,
  pregrading: { score: 1, annotations: [] },
  censoring: {
    scores: [
      { score: 4, shortCode: 'SE3' },
      { score: 3, shortCode: 'SE2' },
      { score: 2, shortCode: 'SE1' },
    ],
    annotations: [],
    nonAnswerDetails: {},
  },
  inspection: { score: 5, shortCodes: ['IN1', 'IN2'] },
}

describe('<QuestionManualScore />', () => {
  let i18n: i18next.i18n

  describe('fi-FI', () => {
    beforeAll(() => {
      i18n = initI18n('fi-FI')
    })

    it('renders without score', () => {
      const props = {
        ...defaultProps,
        scores: undefined,
      }
      expect(renderWithContext(props, []).toJSON()).toMatchSnapshot()
    })

    it('renders with pregrading score', () => {
      const props = {
        ...defaultProps,
        scores: _.pick(defaultScores, 'pregrading', 'answerId', 'questionId'),
      }
      expect(renderWithContext(props, []).toJSON()).toMatchSnapshot()
    })

    it('renders with pregrading and one censor scores', () => {
      const props = {
        ...defaultProps,
        scores: {
          censoring: { scores: [defaultScores.censoring!.scores[0]] },
          ..._.pick(defaultScores, 'pregrading', 'answerId', 'questionId'),
        },
      }
      expect(renderWithContext(props, []).toJSON()).toMatchSnapshot()
    })

    it('renders with pregrading and three censor scores', () => {
      const props = {
        ...defaultProps,
        scores: _.pick(defaultScores, 'pregrading', 'censoring', 'answerId', 'questionId'),
      }
      expect(renderWithContext(props, []).toJSON()).toMatchSnapshot()
    })

    it('renders with pregrading, censor and inspection scores', () => {
      const props = {
        ...defaultProps,
        scores: defaultScores,
      }
      expect(renderWithContext(props, []).toJSON()).toMatchSnapshot()
    })

    it('renders with displayNumber where there are more than one answer', () => {
      const props = {
        ...defaultProps,
        scores: _.pick(defaultScores, 'pregrading', 'answerId', 'questionId'),
      }
      expect(renderWithContext(props, [{} as Element, {} as Element]).toJSON()).toMatchSnapshot()
    })

    it('renders with NoPregrading when there are no scores', () => {
      const props = {
        ...defaultProps,
        scores: _.pick(defaultScores, 'answerId', 'questionId'),
      }
      expect(renderWithContext(props, []).toJSON()).toMatchSnapshot()
    })

    it('renders with NoPregrading when there is pregrading data but no scores', () => {
      const props = {
        ...defaultProps,
        scores: {
          preGrading: { comment: 'comment' },
          ..._.pick(defaultScores, 'answerId', 'questionId'),
        },
      }
      expect(renderWithContext(props, []).toJSON()).toMatchSnapshot()
    })

    it('renders without NoPregrading when pregrading score is 0', () => {
      const props = {
        ...defaultProps,
        scores: {
          pregrading: { score: 0 },
          ..._.pick(defaultScores, 'answerId', 'questionId'),
        },
      }
      expect(renderWithContext(props, []).toJSON()).toMatchSnapshot()
    })

    it('renders without NoPregrading when there are scores', () => {
      const props = {
        ...defaultProps,
        scores: _.pick(defaultScores, 'answerId', 'questionId', 'pregrading', 'censoring'),
      }
      expect(renderWithContext(props, []).toJSON()).toMatchSnapshot()
    })
  })

  describe('sv-FI', () => {
    it('renders with pregrading, censor and inspection scores', () => {
      i18n = initI18n('fi-SV')
      const props = {
        ...defaultProps,
        scores: defaultScores,
      }
      expect(renderWithContext(props, []).toJSON()).toMatchSnapshot()
    })
  })

  const contextProps = {
    answers: [],
    displayNumber: '1',
    hasExternalMaterial: false,
    maxAnswers: 2,
    maxScore: 2,
    level: 2,
    childQuestions: [],
  }

  function renderWithContext(props: QuestionManualScoreProps, answers: Element[]) {
    return create(
      <QuestionContext.Provider value={{ ...contextProps, answers }}>
        <I18nextProvider i18n={i18n}>
          <QuestionManualScore {...props} />
        </I18nextProvider>
      </QuestionContext.Provider>
    )
  }
})
