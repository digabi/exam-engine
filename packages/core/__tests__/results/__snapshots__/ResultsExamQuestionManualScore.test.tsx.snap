// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResultsExamQuestionManualScore /> fi-FI renders with NoPregrading when there are no scores 1`] = `
<div
  className="e-result-scorecount e-float-right e-mrg-b-1"
>
  <span
    className="e-result-scorecount-empty"
  />
   / 6 
   p.
</div>
`;

exports[`<ResultsExamQuestionManualScore /> fi-FI renders with NoPregrading when there is pregrading data but no scores 1`] = `
<div
  className="e-result-scorecount e-float-right e-mrg-b-1"
>
  <span
    className="e-result-scorecount-empty"
  />
   / 6 
   p.
</div>
`;

exports[`<ResultsExamQuestionManualScore /> fi-FI renders with displayNumber where there are more than one answer 1`] = `
<div
  className="e-result-scorecount e-float-right e-mrg-b-1"
>
  <sup
    className="e-result-scorecount-sup e-mrg-r-1"
  >
    1
  </sup>
  <div
    className="e-color-black"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap e-font-size-m"
    >
      <b>
        1
      </b>
       / 6
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      va
    </span>
  </div>
</div>
`;

exports[`<ResultsExamQuestionManualScore /> fi-FI renders with pregrading and one censor scores 1`] = `
<div
  className="e-result-scorecount e-float-right e-mrg-b-1"
>
  <div
    className="e-color-black"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap e-font-size-m"
    >
      <b>
        4
      </b>
       / 6
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      SE3
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      1.s
    </span>
  </div>
  <div
    className="e-color-grey e-font-size-xs"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap"
    >
      1
      
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      va
    </span>
  </div>
</div>
`;

exports[`<ResultsExamQuestionManualScore /> fi-FI renders with pregrading and three censor scores 1`] = `
<div
  className="e-result-scorecount e-float-right e-mrg-b-1"
>
  <div
    className="e-color-black"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap e-font-size-m"
    >
      <b>
        4
      </b>
       / 6
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      SE3
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      3.s
    </span>
  </div>
  <div
    className="e-color-grey e-font-size-xs"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap"
    >
      3
      
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      SE2
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      2.s
    </span>
  </div>
  <div
    className="e-color-grey e-font-size-xs"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap"
    >
      2
      
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      SE1
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      1.s
    </span>
  </div>
  <div
    className="e-color-grey e-font-size-xs"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap"
    >
      1
      
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      va
    </span>
  </div>
</div>
`;

exports[`<ResultsExamQuestionManualScore /> fi-FI renders with pregrading score 1`] = `
<div
  className="e-result-scorecount e-float-right e-mrg-b-1"
>
  <div
    className="e-color-black"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap e-font-size-m"
    >
      <b>
        1
      </b>
       / 6
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      va
    </span>
  </div>
</div>
`;

exports[`<ResultsExamQuestionManualScore /> fi-FI renders with pregrading, censor and inspection scores 1`] = `
<div
  className="e-result-scorecount e-float-right e-mrg-b-1"
>
  <div
    className="e-color-black"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap e-font-size-m"
    >
      <b>
        5
      </b>
       / 6
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      IN1, IN2
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      ta
    </span>
  </div>
  <div
    className="e-color-grey e-font-size-xs"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap"
    >
      4
      
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      SE3
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      3.s
    </span>
  </div>
  <div
    className="e-color-grey e-font-size-xs"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap"
    >
      3
      
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      SE2
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      2.s
    </span>
  </div>
  <div
    className="e-color-grey e-font-size-xs"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap"
    >
      2
      
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      SE1
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      1.s
    </span>
  </div>
  <div
    className="e-color-grey e-font-size-xs"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap"
    >
      1
      
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      va
    </span>
  </div>
</div>
`;

exports[`<ResultsExamQuestionManualScore /> fi-FI renders without NoPregrading when pregrading score is 0 1`] = `
<div
  className="e-result-scorecount e-float-right e-mrg-b-1"
>
  <div
    className="e-color-black"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap e-font-size-m"
    >
      <b>
        0
      </b>
       / 6
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      va
    </span>
  </div>
</div>
`;

exports[`<ResultsExamQuestionManualScore /> fi-FI renders without NoPregrading when there are scores 1`] = `
<div
  className="e-result-scorecount e-float-right e-mrg-b-1"
>
  <div
    className="e-color-black"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap e-font-size-m"
    >
      <b>
        4
      </b>
       / 6
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      SE3
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      3.s
    </span>
  </div>
  <div
    className="e-color-grey e-font-size-xs"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap"
    >
      3
      
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      SE2
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      2.s
    </span>
  </div>
  <div
    className="e-color-grey e-font-size-xs"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap"
    >
      2
      
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      SE1
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      1.s
    </span>
  </div>
  <div
    className="e-color-grey e-font-size-xs"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap"
    >
      1
      
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      va
    </span>
  </div>
</div>
`;

exports[`<ResultsExamQuestionManualScore /> fi-FI renders without score 1`] = `
<div
  className="e-result-scorecount e-float-right e-mrg-b-1"
>
  <span
    className="e-result-scorecount-empty"
  />
   / 6 
   p.
</div>
`;

exports[`<ResultsExamQuestionManualScore /> sv-FI renders with pregrading, censor and inspection scores 1`] = `
<div
  className="e-result-scorecount e-float-right e-mrg-b-1"
>
  <div
    className="e-color-black"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap e-font-size-m"
    >
      <b>
        5
      </b>
       / 6
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      IN1, IN2
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      ta
    </span>
  </div>
  <div
    className="e-color-grey e-font-size-xs"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap"
    >
      4
      
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      SE3
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      3.s
    </span>
  </div>
  <div
    className="e-color-grey e-font-size-xs"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap"
    >
      3
      
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      SE2
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      2.s
    </span>
  </div>
  <div
    className="e-color-grey e-font-size-xs"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap"
    >
      2
      
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      SE1
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      1.s
    </span>
  </div>
  <div
    className="e-color-grey e-font-size-xs"
  >
    <span
      className="e-mrg-r-1 e-nowrap e-nowrap"
    >
      1
      
       p.
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-result-scorecount-shortcode"
    >
      
    </span>
    <span
      className="e-mrg-r-1 e-nowrap e-mrg-r-0"
    >
      va
    </span>
  </div>
</div>
`;
